assets/js/base/context/hooks/cart/use-store-cart-coupons.ts

```diff
@@ -29,103 +29,106 @@ export const useStoreCartCoupons = ( context = '' ): StoreCartCoupon => {
 
-	const results: Pick<
+	const {
+		applyCoupon,
+		removeCoupon,
+		isApplyingCoupon,
+		isRemovingCoupon,
+	}: Pick<
 		StoreCartCoupon,
-		'applyCoupon' | 'removeCoupon' | 'isApplyingCoupon' | 'isRemovingCoupon'
+		| 'applyCoupon'
+		| 'removeCoupon'
+		| 'isApplyingCoupon'
+		| 'isRemovingCoupon'
+		| 'receiveApplyingCoupon'
 	> = useSelect(
 		( select, { dispatch } ) => {
 			const store = select( storeKey );
-			const isApplyingCoupon = store.isApplyingCoupon();
-			const isRemovingCoupon = store.isRemovingCoupon();
-			const {
-				applyCoupon,
-				removeCoupon,
-				receiveApplyingCoupon,
-			}: {
-				applyCoupon: ( coupon: string ) => Promise< boolean >;
-				removeCoupon: ( coupon: string ) => Promise< boolean >;
-				receiveApplyingCoupon: ( coupon: string ) => void;
-			} = dispatch( storeKey );
-
-			const applyCouponWithNotices = ( couponCode: string ) => {
-				applyCoupon( couponCode )
-					.then( ( result ) => {
-						if ( result === true ) {
-							createNotice(
-								'info',
-								sprintf(
-									/* translators: %s coupon code. */
-									__(
-										'Coupon code "%s" has been applied to your cart.',
-										'woo-gutenberg-products-block'
-									),
-									couponCode
-								),
-								{
-									id: 'coupon-form',
-									type: 'snackbar',
-									context,
-								}
-							);
-						}
-					} )
-					.catch( ( error ) => {
-						setValidationErrors( {
-							coupon: {
-								message: decodeEntities( error.message ),
-								hidden: false,
-							},
-						} );
-						// Finished handling the coupon.
-						receiveApplyingCoupon( '' );
-					} );
-			};
-
-			const removeCouponWithNotices = ( couponCode: string ) => {
-				removeCoupon( couponCode )
-					.then( ( result ) => {
-						if ( result === true ) {
-							createNotice(
-								'info',
-								sprintf(
-									/* translators: %s coupon code. */
-									__(
-										'Coupon code "%s" has been removed from your cart.',
-										'woo-gutenberg-products-block'
-									),
-									couponCode
-								),
-								{
-									id: 'coupon-form',
-									type: 'snackbar',
-									context,
-								}
-							);
-						}
-					} )
-					.catch( ( error ) => {
-						createErrorNotice( error.message, {
-							id: 'coupon-form',
-							context,
-						} );
-						// Finished handling the coupon.
-						receiveApplyingCoupon( '' );
-					} );
-			};
+			const actions = dispatch( storeKey );
 
 			return {
-				applyCoupon: applyCouponWithNotices,
-				removeCoupon: removeCouponWithNotices,
-				isApplyingCoupon,
-				isRemovingCoupon,
+				applyCoupon: actions.applyCoupon,
+				removeCoupon: actions.removeCoupon,
+				isApplyingCoupon: store.isApplyingCoupon(),
+				isRemovingCoupon: store.isRemovingCoupon(),
+				receiveApplyingCoupon: actions.receiveApplyingCoupon,
 			};
 		},
 		[ createErrorNotice, createNotice ]
 	);
 
+	const applyCouponWithNotices = ( couponCode: string ) => {
```

Comment:
I don't know how often `createNotice` or `setValidationErrors` or `createErrorNotice` change (can't remember if they themselves are dispatch function) but one optimization here might be to memoize the new composite dispatchers via `useCallback`. I _think_ that might have been why they were in the `useSelect` callback originally so that we memoized them based on the dependencies.